crates/engine/tree/src/tree/payload_processor/mod.rs

```diff
@@ -318,6 +324,144 @@ where
+    fn dispatch_trie_spawn<BPF>(
+        &self,
+        configured_trie: ConfiguredSparseTrie,
+        sparse_trie_rx: mpsc::Receiver<SparseTrieUpdate>,
+        proof_task_handle: BPF,
+        state_root_tx: mpsc::Sender<
+            Result<StateRootComputeOutcome<ConfiguredSparseTrie>, ParallelStateRootError>,
+        >,
+        is_revealed: bool,
+    ) where
+        BPF: BlindedProviderFactory + Clone + Send + Sync + 'static,
+        BPF::AccountNodeProvider: BlindedProvider + Send + Sync,
+        BPF::StorageNodeProvider: BlindedProvider + Send + Sync,
+    {
+        match configured_trie {
+            ConfiguredSparseTrie::Serial(boxed_serial) => {
+                let trie = if is_revealed {
+                    Some(SparseTrie::Revealed(boxed_serial))
+                } else {
+                    Some(SparseTrie::Blind(Some(boxed_serial)))
+                };
+                self.spawn_trie_task(sparse_trie_rx, proof_task_handle, state_root_tx, trie);
+            }
+            ConfiguredSparseTrie::Parallel(boxed_parallel) => {
+                let trie = if is_revealed {
+                    Some(SparseTrie::Revealed(boxed_parallel))
+                } else {
+                    Some(SparseTrie::Blind(Some(boxed_parallel)))
+                };
+                self.spawn_trie_task(sparse_trie_rx, proof_task_handle, state_root_tx, trie);
+            }
+        }
+    }
+
+    /// Helper method that handles sparse trie task spawning.
+    ///
+    /// If we have a stored trie, we will re-use it for spawning. If we do not have a stored trie,
+    /// we will create a new trie based on the configured trie type (parallel or serial).
+    fn spawn_sparse_trie_task<BPF>(
+        &self,
+        sparse_trie_rx: mpsc::Receiver<SparseTrieUpdate>,
+        proof_task_handle: BPF,
+        state_root_tx: mpsc::Sender<
+            Result<StateRootComputeOutcome<ConfiguredSparseTrie>, ParallelStateRootError>,
+        >,
+        stored_trie: Option<SparseTrie<ConfiguredSparseTrie>>,
+        use_parallel_for_new: bool,
+    ) where
+        BPF: BlindedProviderFactory + Clone + Send + Sync + 'static,
+        BPF::AccountNodeProvider: BlindedProvider + Send + Sync,
+        BPF::StorageNodeProvider: BlindedProvider + Send + Sync,
+    {
+        match stored_trie {
+            Some(SparseTrie::Revealed(boxed)) => {
+                self.dispatch_trie_spawn(
+                    *boxed,
+                    sparse_trie_rx,
+                    proof_task_handle,
+                    state_root_tx,
+                    // this is revealed
+                    true,
+                );
+            }
+            Some(SparseTrie::Blind(Some(boxed))) => {
+                self.dispatch_trie_spawn(
+                    *boxed,
+                    sparse_trie_rx,
+                    proof_task_handle,
+                    state_root_tx,
+                    // this is blind
+                    false,
+                );
```

Comment:
can probably be merged into into match arm with `stored_trie.is_revealed()` as last argument